File: 1 - metodosBusca/pacoteBusca_JomiHubner/src/busca/BuscaHeuristica.java
package busca;

import java.util.Comparator;

/**
 *   Algoritmos de Busca heurística
 *
 *   @author Jomi Fred Hübner
 */
public abstract class BuscaHeuristica extends Busca {
    
    /** busca sem mostrar status */
    public BuscaHeuristica() {
    }
    
    /** busca mostrando status */
    public BuscaHeuristica(MostraStatusConsole ms) {
        super(ms);
    }

    /** comparador para ordenar os nodos por F */
    Comparator<Nodo> getNodoComparatorF() {
        return new Comparator<Nodo>() {
            public int compare(Nodo no1, Nodo no2) {
                try {
                    //Heuristica eo1 = (Heuristica)no1.estado;
                    //Heuristica eo2 = (Heuristica)no2.estado;
                    int f1 = no1.f();
                    int f2 = no2.f();
                    if (f1 > f2) {
                        return 1;
                    } else if (f1 == f2) {
                        return 0; 
                    } else {
                        return -1;
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
                return 0;
            }
        };
    }

    /** comparador para ordenar os Estados por h */
    /*<Heuristica> getEstadoComparatorH() {
        return new Comparator<Heuristica>() {
            public int compare(Heuristica eo1, Heuristica eo2) {
                try {
                	//Heuristica eo1 = (Heuristica)o1;
                    //Heuristica eo2 = (Heuristica)o2;
                    int h1 = eo1.h();
                    int h2 = eo2.h();
                    if (h1 > h2) {
                        return 1;
                    } else if (h1 == h2) {
                        return 0; 
                    } else {
                        return -1;
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
                return 0;
            }
        };
    }
    */
}

File: 1 - metodosBusca/python/app/core/busca/busca_heuristica.py
from abc import ABC
from functools import cmp_to_key
from app.core.busca.busca import Busca

class BuscaHeuristica(Busca, ABC):
    def __init__(self, mostra_status_console=None):
        super().__init__(mostra_status_console)

    def get_nodo_comparator_f(self):
        def comparator(no1, no2):
            try:
                f1 = no1.f()
                f2 = no2.f()
                if f1 > f2:
                    return 1
                elif f1 == f2:
                    return 0
                else:
                    return -1
            except Exception as e:
                print(e)
                return 0

        return cmp_to_key(comparator)

    # Uncomment and implement if needed
    # def get_estado_comparator_h(self):
    #     def comparator(eo1, eo2):
    #         try:
    #             h1 = eo1.h()
    #             h2 = eo2.h()
    #             if h1 > h2:
    #                 return 1
    #             elif h1 == h2:
    #                 return 0
    #             else:
    #                 return -1
    #         except Exception as e:
    #             print(e)
    #             return 0
    #
    #     return cmp_to_key(comparator)